synapse/lib/lmdb.py

```diff
@@ -553,3 +565,411 @@ def encodeValAsKey(v, isprefix=False):
             return _STR_VAL_MARKER + v_enc
         else:
             return _STR_VAL_MARKER + v_enc + b'\x00'
+
+class Slab(s_base.Base):
+    '''
+    A "monolithic" LMDB instance for use in a asyncio loop thread.
+    '''
+    COMMIT_PERIOD = 1.0  # time between commits
+
+    async def __anit__(self, path, **opts):
+        await s_base.Base.__anit__(self)
+
+        self.path = path
+        self.optspath = os.path.join(path, 'opts.json')
+
+        if os.path.isfile(self.optspath):
+            opts.update(s_common.jsload(self.optspath))
+
+        self.mapsize = opts.get('map_size')
+        if self.mapsize is None:
+            raise Exception('Slab requires map_size!')
+
+        # save the transaction deltas in case of error...
+        self.xactops = []
+        self.recovering = False
+
+        # if growsize is not set, we double...
```

Comment:
Is the order of operations with the .pop and setdefault calls correct?

Reply:
We have three different sets of values values here - runtime, ondisk, and default values all being smooshe together here; all within the same namespace which is then used to  configure ourselves AND create a lmdb environment.   It's a bit to untangle what exactly is going on here. It dovetails into the previous note about config helpers.

Reply:
ok